contracts/external/traders/CanonicalOrders.sol

```diff
@@ -0,0 +1,842 @@
+            FILE,
+            "Cannot approve canceled order",
+            orderHash
+        );
+        g_status[orderHash] = OrderStatus.Approved;
+        emit LogCanonicalOrderApproved(
+            orderHash,
+            approver,
+            order.baseMarket,
+            order.quoteMarket
+        );
+    }
+
+    // ============ Private Helper Functions ============
+
+    /**
+     * Verifies that the order is still fillable for the particular accounts and markets specified.
+     */
+    function verifyOrderAndAccountsAndMarkets(
+        OrderInfo memory orderInfo,
+        Account.Info memory makerAccount,
+        Account.Info memory takerAccount,
+        uint256 inputMarketId,
+        uint256 outputMarketId,
+        Types.Wei memory inputWei
+    )
+        private
+        view
+    {
+        // verify triggerPrice
+        if (orderInfo.order.triggerPrice > 0) {
+            uint256 currentPrice = getCurrentPrice(
+                orderInfo.order.baseMarket,
+                orderInfo.order.quoteMarket
+            );
+            Require.that(
+                isBuy(orderInfo.order)
+                    ? currentPrice >= orderInfo.order.triggerPrice
+                    : currentPrice <= orderInfo.order.triggerPrice,
+                FILE,
+                "Order triggerPrice not triggered",
+                currentPrice
+            );
+        }
+
+        // verify expriy
+        Require.that(
+            orderInfo.order.expiration == 0 || orderInfo.order.expiration >= block.timestamp,
+            FILE,
+            "Order expired",
+            orderInfo.orderHash
+        );
+
+        // verify maker
+        Require.that(
+            makerAccount.owner == orderInfo.order.makerAccountOwner &&
+            makerAccount.number == orderInfo.order.makerAccountNumber,
+            FILE,
+            "Order maker account mismatch",
+            orderInfo.orderHash
+        );
+
+        // verify taker
+        Require.that(
+            orderInfo.order.taker == address(0) || orderInfo.order.taker == takerAccount.owner,
+            FILE,
+            "Order taker mismatch",
+            orderInfo.orderHash
+        );
+
+        // verify markets
+        Require.that(
+            (
+                orderInfo.order.baseMarket == outputMarketId &&
+                orderInfo.order.quoteMarket == inputMarketId
+            ) || (
+                orderInfo.order.quoteMarket == outputMarketId &&
+                orderInfo.order.baseMarket == inputMarketId
+            ),
+            FILE,
+            "Market mismatch",
+            orderInfo.orderHash
+        );
+
+        // verify inputWei
+        Require.that(
+            !inputWei.isZero(),
+            FILE,
+            "InputWei is zero",
+            orderInfo.orderHash
+        );
+        Require.that(
+            inputWei.sign ==
+                ((orderInfo.order.baseMarket == inputMarketId) == isBuy(orderInfo.order)),
```

Comment:
nit: this is really confusing to parse

Maybe at least put some comments in as to what this means, I can't really tell if it's correct

Reply:
Added comment